lib/IMAP/Charset/Converter.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/*
+ * *
+ *  *
+ *  * @copyright 2023 Anna Larch <anna.larch@gmx.net>
+ *  *
+ *  * @author Anna Larch <anna.larch@gmx.net>
+ *  *
+ *  * This library is free software; you can redistribute it and/or
+ *  * modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+ *  * License as published by the Free Software Foundation; either
+ *  * version 3 of the License, or any later version.
+ *  *
+ *  * This library is distributed in the hope that it will be useful,
+ *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+ *  *
+ *  * You should have received a copy of the GNU Affero General Public
+ *  * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *  *
+ *
+ */
+
+namespace OCA\Mail\IMAP\Charset;
+
+use Horde_Mime_Part;
+use OCA\Mail\Exception\ServiceException;
+
+class Converter {
+
+	/**
+	 * @param Horde_Mime_Part $p
+	 * @return string
+	 * @throws ServiceException
+	 */
+	public function convert(Horde_Mime_Part $p): string {
+		$data = $p->getContents();
+		if ($data === null) {
+			return '';
+		}
+
+		// Only convert encoding if it is explicitly specified in the header because text/calendar
+		// data is utf-8 by default.
+		$charset = $p->getCharset();
+		if($charset !== null && strtoupper($charset) === 'UTF-8') {
+			return $data;
+		}
+
+		// No charset specified
+		$charset = mb_detect_encoding($data, 'UTF-8', true);
```

Comment:
it's because the part's charset is overwritten here except for the case of UTF-8. "No charset specified" is possibly wrong here. The charset might be set but differ from "utf-8".